doc/Variables.md

```diff
@@ -0,0 +1,117 @@
+   possible, it is preferred to reused existing messages, even if they are not a perfect fit. Similarly, because the
+   fuse Variables type must match exactly to be combined together, it is always better to reuse existing Variable
+   types.
+
+For the most part, reasonable variable types will be fairly obvious: 2D position, 2D velocity, 2D acceleration, etc.
+The biggest debate is generally whether to include the linear and angular information into a single variable (e.g.
+a 2D pose consists of a 2D position and 2D orientation), or the if they should be separate. For fuse, it was decided
+to keep the linear and angular components separate. The [`fuse_variables`](fuse_variables) package provides a set
+of common, reusable 2D and 3D variable types. And submissions of new variable types are always welcome.
+
+## Variable API
+
+Like basically everything in fuse, the Variable system is designed to be extensible. The
+[`fuse_core::Variable`](fuse_core/include/fuse_core/variable.h) base class defines the minimum interface required
+for all derived variable types.
+
+* `Derived::type() -> std::string`: All derived Variables must implement a `type()` method that returns the
+  fully-qualified class name. This is of the form `namespace::ClassName`. Under most circumstances, the base class
+  implementation will return the correct class name. However, templated Variable classes may need to override the
+  `type()` method to return the desired name.
+* `Derived::size() -> size_t`: The derived Variable type must return the number of dimensions of the variable. This
+  will likely just return a fixed constant.
+* `Derived::data() -> double*`: The _value_ portion of the derived Variable must be accessible from a contiguous memory
+  location of size `derived.size() * sizeof(double)`.
+* `Derived::uuid() -> fuse_core::UUID`: This is the _identity_ of the variable. Each derived class is required to
+  return a unique ID to act as the identity of the variable. The UUID must be different for each different combination
+  of class and _identity_ property values. In the robotics example, the 2D pose UUID should be built from the 2D pose
+  type string, the `robot_id` field, and the `timestamp` field. Importantly, if a Variable instance has the same
+  _identity_ property values, **the same UUID should be returned**. This is how multiple sensors can produce the same
+  variable identity in a distributed fashion. Again, in the robotics example, the UUID of a 2D pose with a `robot_id`
+  of 'R1' and a `timestamp` of 10.0 produced by the odometry sensor should be the same UUID of a 2D pose with a
+  `robot_id` of 'R1' and a `timestamp` of 10.0 produced by the laserscan sensor. Some functions for generating UUIDs
+  are provided [here](fuse_core/include/fuse_core/uuid.h).
+* `Derived::print(std::ostream& stream)`: It's nice to print variable information during debugging. Each
```

Comment:
Given that we have the UUID, the data, and the data length, it makes me wonder if having a base class print implementation might be worthwhile.

Reply:
Probably. Making a separate ticket.